src/main/java/io/cdap/plugin/gcp/bigquery/connector/BigQueryConnector.java

```diff
@@ -233,19 +242,68 @@ private BigQuery getBigQuery(String project) throws IOException {
   }
 
   private List<StructuredRecord> getTableData(BigQuery bigQuery, String datasetProject, String dataset, String table,
-    int limit)
+    int limit, String sampleType, @Nullable String strata, @Nullable Long timeoutMs)
     throws IOException {
-    String query =
-      String.format("SELECT * FROM `%s.%s.%s` LIMIT %d", datasetProject, dataset, table, limit);
+    if (sampleType == null) {
+      sampleType = "first";
+    }
+    String tableName = String.format("`%s.%s.%s`", datasetProject, dataset, table);
+    String query;
+    switch (SampleType.fromString(sampleType)) {
+      case RANDOM:
+        query = String.format("WITH table AS (\n" +
+                                "  SELECT *, RAND() AS r\n" +
+                                "  FROM %s\n" +
+                                "  WHERE RAND() < 2*%d/(SELECT COUNT(*) FROM %s)\n" +
+                                ")\n" +
+                                "SELECT * EXCEPT (r)\n" +
+                                "FROM table\n" +
+                                "ORDER BY r\n" +
+                                "LIMIT %d", tableName, limit, tableName, limit);
+        break;
+      case STRATIFIED:
+        if (strata == null) {
+          throw new IllegalArgumentException("No strata column given.");
+        }
+        query = String.format("SELECT * EXCEPT (`seqnum`, `cnt`)\n" +
+                        "FROM (\n" +
+                        "SELECT *, row_number() OVER (ORDER BY %s, RAND()) AS seqnum,\n" +
+                        "COUNT(*) OVER () as cnt,\n" +
+                        "FROM %s\n" +
+                        ") %s\n" +
+                        "WHERE MOD(seqnum, CAST(cnt / %d AS INT64)) = 1\n" +
+                        "ORDER BY %s\n" +
+                        "LIMIT %d",
+                strata, tableName, tableName, limit, strata, limit);
+        break;
+      default:
+        query =
+          String.format("SELECT * FROM %s LIMIT %d", tableName, limit);
+    }
+
     QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query).build();
```

Comment:
I see a `setTimeoutMs` method in the `QueryJobConfiguration.Builder` class. 

https://javadoc.io/static/com.google.cloud/google-cloud-bigquery/1.137.1/com/google/cloud/bigquery/QueryJobConfiguration.Builder.html#setJobTimeoutMs-java.lang.Long-

Take a look at this method and see if it might be useful in this case.

Reply:
I gave this method a try at first, but it turns out that it doesn't guarantee a timeout--in the documentation it says "BigQuery _may_ attempt to terminate the job." And when I was playing around with it, it seemed like most of the time it ran significantly longer than the inputted timeout, which would be unexpected to the user.

Reply:
I see, then disregard my comment.